### PR TITLE
[RDY] More work on highlight tests + update documentation

### PR DESCRIPTION
**Pull request by bfredl**
Continuation of #1835.
- Better error messages for unexpected highlights
- Use strict mode per default and in existing tests. It is still possible to ignore unexpected highlights by
  setting `ignore = true`
- Update documentation

I will "dog-food" this and write some tests for search highlighting and such, to uncover remaining rough corners. 

There is an api function that translates color names to rgb values, but when using the `snapshot_util` workflow, it would be convenient to be able to do the reverse, i.e. given a present rgb value, find the canonical color name for it. 
One possibility: when initializing the `screen`, populate a table with all color name / rgb value pairs. Then, also, every specific test need not use `nvim('name_to_color', 'MyColor')`, instead `screen.colors.MyColor`.

### CODE REVIEW

**Comment by justinmk** [comment]
Nice!

**Comment by justinmk** [comment]
> Use strict mode per default 

Do you mean lua strict mode or something else? Last I looked, I could not find an "official" lua strict mode, just some scripts in penlight or whatever.

**Comment by bfredl** [comment]
Sorry, bad term, I mean the stricter way of interpreting highlight tests as
discussed in #1835 (unexpected hightlight groups are an error, unless
explicitly ignored)
Den 24 jan 2015 17:47 skrev "Justin M. Keyes" <notifications@github.com>:

File: test/functional/ui/screen.lua [hunk 1]
```diff
@@ -54,12 +54,17 @@
 -- having the exact same set of attributes will be substituted by "{K:S}",
 -- where K is a key associated the attribute set via the second argument of
 -- "expect".
+-- If a transformation table is present, unexpected attribute sets in the final
+-- state is considered an error. To make testing simpler, a list of attribute sets that
+-- should be ignored can be passed as a third argument. Alternatively, this third
+-- argument can be "true" to indicate that all unexpected attribute sets should be ignored.
```
Comment (oni-link) [inline test/functional/ui/screen.lua]: Do we also use a maximum of 80 columns for lua files? `:set colorcolumn=80` shows some
lines longer than that.

**Comment by tarruda** [comment]
Great work @bfredl, thank you for this.

> There is an api function that translates color names to rgb values, but when using the snapshot_util workflow, it would be convenient to be able to do the reverse, i.e. given a present rgb value, find the canonical color name for it. 
One possibility: when initializing the screen, populate a table with all color name / rgb value pairs. Then, also, every specific test need not use nvim('name_to_color', 'MyColor'), instead screen.colors.MyColor.

Eventually I'd like to remove this color name->rgb translation from nvim and let `:highlight` associate arbitrary keys/values with groups.

File: test/functional/ui/screen.lua (hunk 1 above)
Comment (bfredl) [inline test/functional/ui/screen.lua]: :+1: I will cleanup formatting before this PR gets closer to ready.

**Comment by bfredl** [comment]
@tarruda How will the default highlightings (used in tests) look then? both rgb values and color names?

**Comment by tarruda** [comment]
> @tarruda How will the default highlightings (used in tests) look then? both rgb values and color names?

Yes. When a highlight group contains `guifg=yellow guibg="#ffffff"`, the `{'guifg': 'yellow', 'guibg': '#ffffff'}` dict will be passed unmodified to the UI layer. In other word,s the highlight keys/values only have meaning in the UI.

We can also optimize to pass the dict only the first time a highlight group is used, and use a unique identifier in subsequent highlight changes. That way the UI can store the attributes in whatever format is more optimal for rendering.

**Comment by bfredl** [comment]
So then, color names (as used exclusively in the `default` theme) should be translated on the gui side? In that case, I still think it should be possible to retrieve the entire table, rather than just a single name a time. (As clients probably would cache this information, anyway, like the tests already do)
Caching the attribute sets also sounds like a worthwhile optimization, I think this might also make the screen test helpers simpler. :+1: 
Anyway, in this PR I will do what's most convenient for the moment, the tests could easily be updated when the highlight code is refactored later.

**Comment by bfredl** [comment]
Now `snapshot_util` is a bit nicer about colors: instead of just spitting out a rgb-value in decimal encoding, it uses the first name (unique?) found in the name->rgb table.

**Comment by bfredl** [comment]
Fixed the documentation formatting. Marking RFC. Use [this link](https://github.com/neovim/neovim/pull/1883/files?w=1) to verify that I haven't changed the values in the colortable in `syntax.c` (I resisted the temptation to factor `highlight.c` out of `syntax.c` in the same PR, I mean, literally, there is a marker in the middle of `syntax.c` that says that the syntax code ends and the general highlighting code starts....)

File: src/nvim/api/vim.c [hunk 2]
```diff
@@ -552,6 +552,18 @@ Integer vim_name_to_color(String name)
   return name_to_color((uint8_t *)name.data);
 }
 
+Dictionary vim_get_color_table(void)
```
Comment (justinmk) [inline src/nvim/api/vim.c]: `vim_get_color_map` ? We should probably stick to consistently referring to dictionaries/maps for the public API at least. No need to introduce another synonym "table".

Comment (bfredl) [inline src/nvim/api/vim.c]: Good point :+1:

**Comment by tarruda** [comment]
LGTM

**Comment by bfredl** [comment]
Rebased.

**Comment by tarruda** [comment]
Lets hold on this PR for a bit to see if #1572 is fully fixed and merged :)

**Comment by bfredl** [comment]
I used this quite conveniently to generate some basic `hlsearch` tests (will push PR after this), so marking as RDY if no one has more objections/suggestions.

**Comment by tarruda** [comment]
:+1: merged, thanks

**Pull request closed without merging**